File: implementations/ipfs/async-resolver/src/wrap/imported/uri_resolver_module/serialization.rs
use serde::{Serialize, Deserialize};
use polywrap_wasm_rs::{
    BigInt,
    BigNumber,
    Map,
    Context,
    DecodeError,
    EncodeError,
    Read,
    ReadDecoder,
    Write,
    WriteEncoder,
    JSON,
};

use crate::UriResolverMaybeUriOrManifest;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsTryResolveUri {
    pub authority: String,
    pub path: String,
}

pub fn deserialize_try_resolve_uri_args(args: &[u8]) -> Result<ArgsTryResolveUri, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing imported module-type: try_resolve_uri Args".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _authority: String = String::new();
    let mut _authority_set = false;
    let mut _path: String = String::new();
    let mut _path_set = false;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "authority" => {
                reader.context().push(&field, "String", "type found, reading argument");
                _authority = reader.read_string()?;
                _authority_set = true;
                reader.context().pop();
            }
            "path" => {
                reader.context().push(&field, "String", "type found, reading argument");
                _path = reader.read_string()?;
                _path_set = true;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_authority_set {
        return Err(DecodeError::MissingField("authority: String.".to_string()));
    }
    if !_path_set {
        return Err(DecodeError::MissingField("path: String.".to_string()));
    }

    Ok(ArgsTryResolveUri {
        authority: _authority,
        path: _path,
    })
}

pub fn serialize_try_resolve_uri_args(args: &ArgsTryResolveUri) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) imported module-type: try_resolve_uri Args".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_try_resolve_uri_args(args, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_try_resolve_uri_args<W: Write>(args: &ArgsTryResolveUri, writer: &mut W) -> Result<(), EncodeError> {
    writer.write_map_length(&2)?;
    writer.context().push("authority", "String", "writing property");
    writer.write_string("authority")?;
    writer.write_string(&args.authority)?;
    writer.context().pop();
    writer.context().push("path", "String", "writing property");
    writer.write_string("path")?;
    writer.write_string(&args.path)?;
    writer.context().pop();
    Ok(())
}

pub fn serialize_try_resolve_uri_result(result: &Option<UriResolverMaybeUriOrManifest>) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) imported module-type: try_resolve_uri Result".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_try_resolve_uri_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_try_resolve_uri_result<W: Write>(result: &Option<UriResolverMaybeUriOrManifest>, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("tryResolveUri", "Option<UriResolverMaybeUriOrManifest>", "writing result");
    if result.is_some() {
        UriResolverMaybeUriOrManifest::write(result.as_ref().unwrap(), writer)?;
    } else {
        writer.write_nil()?;
    }
    writer.context().pop();
    Ok(())
}

pub fn deserialize_try_resolve_uri_result(result: &[u8]) -> Result<Option<UriResolverMaybeUriOrManifest>, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing imported module-type: try_resolve_uri Result".to_string();
    let mut reader = ReadDecoder::new(result, context);

    reader.context().push("tryResolveUri", "Option<UriResolverMaybeUriOrManifest>", "reading function output");
    let mut object: Option<UriResolverMaybeUriOrManifest> = None;
    if !reader.is_next_nil()? {
        object = Some(UriResolverMaybeUriOrManifest::read(&mut reader)?);
    } else {
        object = None;
    }
    let res = object;
    reader.context().pop();
    Ok(res)
}

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ArgsGetFile {
    pub path: String,
}

pub fn deserialize_get_file_args(args: &[u8]) -> Result<ArgsGetFile, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing imported module-type: get_file Args".to_string();

    let mut reader = ReadDecoder::new(args, context);
    let mut num_of_fields = reader.read_map_length()?;

    let mut _path: String = String::new();
    let mut _path_set = false;

    while num_of_fields > 0 {
        num_of_fields -= 1;
        let field = reader.read_string()?;

        match field.as_str() {
            "path" => {
                reader.context().push(&field, "String", "type found, reading argument");
                _path = reader.read_string()?;
                _path_set = true;
                reader.context().pop();
            }
            err => return Err(DecodeError::UnknownFieldName(err.to_string())),
        }
    }
    if !_path_set {
        return Err(DecodeError::MissingField("path: String.".to_string()));
    }

    Ok(ArgsGetFile {
        path: _path,
    })
}

pub fn serialize_get_file_args(args: &ArgsGetFile) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) imported module-type: get_file Args".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_get_file_args(args, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_get_file_args<W: Write>(args: &ArgsGetFile, writer: &mut W) -> Result<(), EncodeError> {
    writer.write_map_length(&1)?;
    writer.context().push("path", "String", "writing property");
    writer.write_string("path")?;
    writer.write_string(&args.path)?;
    writer.context().pop();
    Ok(())
}

pub fn serialize_get_file_result(result: &Option<Vec<u8>>) -> Result<Vec<u8>, EncodeError> {
    let mut encoder_context = Context::new();
    encoder_context.description = "Serializing (encoding) imported module-type: get_file Result".to_string();
    let mut encoder = WriteEncoder::new(&[], encoder_context);
    write_get_file_result(result, &mut encoder)?;
    Ok(encoder.get_buffer())
}

pub fn write_get_file_result<W: Write>(result: &Option<Vec<u8>>, writer: &mut W) -> Result<(), EncodeError> {
    writer.context().push("getFile", "Option<Vec<u8>>", "writing result");
    writer.write_optional_bytes(result)?;
    writer.context().pop();
    Ok(())
}

pub fn deserialize_get_file_result(result: &[u8]) -> Result<Option<Vec<u8>>, DecodeError> {
    let mut context = Context::new();
    context.description = "Deserializing imported module-type: get_file Result".to_string();
    let mut reader = ReadDecoder::new(result, context);

    reader.context().push("getFile", "Option<Vec<u8>>", "reading function output");
    let res = reader.read_optional_bytes()?;
    reader.context().pop();
    Ok(res)
}

File: implementations/ipfs/sync-resolver/src/__tests__/simple-wrapper/src/wrap/Module/serialization.ts
import {
  Read,
  ReadDecoder,
  Write,
  WriteSizer,
  WriteEncoder,
  Box,
  BigInt,
  BigNumber,
  JSON,
  Context
} from "@polywrap/wasm-as";
import * as Types from "..";

export class Args_simpleMethod {
  arg: string;
}

export function deserializesimpleMethodArgs(argsBuf: ArrayBuffer): Args_simpleMethod {
  const context: Context = new Context("Deserializing module-type: simpleMethod Args");
  const reader = new ReadDecoder(argsBuf, context);
  let numFields = reader.readMapLength();

  let _arg: string = "";
  let _argSet: bool = false;

  while (numFields > 0) {
    numFields--;
    const field = reader.readString();

    reader.context().push(field, "unknown", "searching for property type");
    if (field == "arg") {
      reader.context().push(field, "string", "type found, reading property");
      _arg = reader.readString();
      _argSet = true;
      reader.context().pop();
    }
    reader.context().pop();
  }

  if (!_argSet) {
    throw new Error(reader.context().printWithContext("Missing required argument: 'arg: String'"));
  }

  return {
    arg: _arg
  };
}

export function serializesimpleMethodArgs(args: Args_simpleMethod): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing) module-type: simpleMethod Args");
  const sizer = new WriteSizer(sizerContext);
  writesimpleMethodArgs(sizer, args);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) module-type: simpleMethod Args");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writesimpleMethodArgs(encoder, args);
  return buffer;
}

export function writesimpleMethodArgs(
  writer: Write,
  args: Args_simpleMethod
): void {
  writer.writeMapLength(1);
  writer.context().push("arg", "string", "writing property");
  writer.writeString("arg");
  writer.writeString(args.arg);
  writer.context().pop();
}

export function serializesimpleMethodResult(result: string): ArrayBuffer {
  const sizerContext: Context = new Context("Serializing (sizing) module-type: simpleMethod Result");
  const sizer = new WriteSizer(sizerContext);
  writesimpleMethodResult(sizer, result);
  const buffer = new ArrayBuffer(sizer.length);
  const encoderContext: Context = new Context("Serializing (encoding) module-type: simpleMethod Result");
  const encoder = new WriteEncoder(buffer, sizer, encoderContext);
  writesimpleMethodResult(encoder, result);
  return buffer;
}

export function writesimpleMethodResult(writer: Write, result: string): void {
  writer.context().push("simpleMethod", "string", "writing property");
  writer.writeString(result);
  writer.context().pop();
}

export function deserializesimpleMethodResult(buffer: ArrayBuffer): string {
  const context: Context = new Context("Deserializing module-type: simpleMethod Result");
  const reader = new ReadDecoder(buffer, context);

  reader.context().push("simpleMethod", "string", "reading function output");
  const res: string = reader.readString();
  reader.context().pop();

  return res;
}
